effect/jvm/src/main/scala/scalaz/effect/SafeApp.scala

```diff
@@ -19,16 +24,23 @@ package scalaz.effect
  * }}}
  */
 trait SafeApp extends RTS {
-  type Error
+
+  final class ExitStatus(val code: Int)
```

Comment:
This is good although it doesn't address the fundamental question:

*What should happen when the main fiber terminates?*

The answers to this question can be:

1. Exit immediately.
2. Exit when no thread in the RTS has any work.
3. Do not exit under any circumstance.

This may be tricky to implement, so I'd suggest to make the API now, and I can help implement (2). The other two can be implemented now and behavior of (2) can simply replicate behavior of (3).

Reply:
As for API, perhaps `ExitStatus` can include `ResidentStatus` which can be one of `ExitNow`, `ExitWhenDone`, `DoNotExit`. 

Then you'd have: `ExitStatus(23, ExitNow)`, for example.

Or perhaps others have suggestions for a more attractive API.

Actually, it occurs to me that `DoNotExit` is incompatible with an exit code, so likely `ResidentStatus` should be folded into `ExitStatus` to give something like:

```scala
sealed trait ExitStatus
object ExitStatus {
  case object ExitNow(code: Int) extends ExitStatus
  case object ExitWhenDone(code: Int) extends ExitStatus
  case object DoNotExit extends ExitStatus
}
```

Reply:
Hey @jdegoes thanks for the quick response.

This PR was meant to solve just #1759 (Change SafeApp to expect an error free main function) and I think you are talking more about #1760 (Change SafeApp to support MainResult), I guess my "+ introducing ExitStatus" was confusing, it was just for carrying the status the jvm should exit with.

But don't worry! I'll include this too and solve both issues in this PR.

Reply:
Two for one, I'll take it. 😉

Reply:
@jdegoes Done. Now with `ExitStatus` you can define when to exit.